src/ot/layout/mod.rs

```diff
@@ -0,0 +1,740 @@
+//! OpenType layout.
+
+mod apply;
+mod common;
+mod context_lookups;
+mod gpos;
+mod gsub;
+mod kern;
+mod matching;
+
+use std::convert::TryFrom;
+use std::slice;
+
+use ttf_parser::parser::{NumFrom, Offset, Offset16, Stream};
+use ttf_parser::GlyphId;
+
+use super::{Map, ShapePlan, TableIndex};
+use crate::buffer::Buffer;
+use crate::common::TagExt;
+use crate::{ffi, Font, Tag};
+use apply::{Apply, ApplyContext, WouldApply, WouldApplyContext};
+use common::{FeatureIndex, LangIndex, LookupIndex, ScriptIndex, SubstPosTable, VariationIndex};
+use gpos::PosTable;
+use gsub::SubstTable;
+
+pub const MAX_NESTING_LEVEL: usize = 6;
+pub const MAX_CONTEXT_LENGTH: usize = 64;
+
+pub const SCRIPT_NOT_FOUND_INDEX: u32 = 0xFFFF;
+pub const LANGUAGE_NOT_FOUND_INDEX: u32 = 0xFFFF;
+pub const FEATURE_NOT_FOUND_INDEX: u32 = 0xFFFF;
+pub const FEATURE_VARIATION_NOT_FOUND_INDEX: u32 = 0xFFFFFFFF;
+
+/// A lookup-based layout table (GSUB or GPOS).
+pub trait LayoutTable {
+    /// The tag of this table.
+    const TAG: Tag;
+
+    /// The index of this table.
+    const INDEX: TableIndex;
+
+    /// Whether lookups in this table can be applied to the buffer in-place.
+    const IN_PLACE: bool;
+
+    /// The kind of lookup stored in this table.
+    type Lookup: LayoutLookup;
+
+    /// Get the lookup at the specified index.
+    fn get_lookup(&self, index: LookupIndex) -> Option<Self::Lookup>;
+}
+
+/// A lookup in a layout table.
+pub trait LayoutLookup: Apply {
+    /// The lookup's lookup_props.
+    fn props(&self) -> u32;
+
+    /// Whether the lookup has to be applied backwards.
+    fn is_reverse(&self) -> bool;
+}
+
+// GDEF
+// Note: GDEF blocklisting was removed for now because we use
+//       ttf_parser's GDEF parsing routines.
+
+/// rb_ot_layout_has_glyph_classes:
+/// @face: #rb_face_t to work upon
+///
+/// Tests whether a face has any glyph classes defined in its GDEF table.
+///
+/// Return value: true if data found, false otherwise
+#[no_mangle]
+pub extern "C" fn rb_ot_layout_has_glyph_classes(face: *const ffi::rb_face_t) -> ffi::rb_bool_t {
+    // TODO: Find out through ttfp_face when that's reachable.
+    (|| {
```

Comment:
Please use a separate Rust function that returns Option instead.

`extern` functions, in general, should not have any logic. Just mapping between C++ and Rust types.

Reply:
I did that for most functions, but not for this one because I don't really think it should exist as a Rust function ever. It's just a temporary workaround because I couldn't access `ttfp_face::Face::has_glyph_classes` there. But I don't really mind, I can wrap it in a function.

Reply:
Ok, in this case you can leave it as is.